kernel/rpi.c

```diff
@@ -125,29 +159,33 @@ intptr_t platform_stack(void) {
 
 __noreturn__ void platform_init(void *atags) {
   (void)atags;
+
+  /* shields up, weapons armed - going live */
+  kernel_entry(0, 0, 0);
+
   for (;;) {
     /* --- DO NOTHING */
   }
 }
 
-__init__ long arm64_init(void) {
+__init__ struct addr_pair arm64_init(void) {
   uint64_t x;
   long cpu = get_cpu();
   enable_cache();
   invalidate_tlb();
 
+  struct addr_pair result = {.pc = 0, .sp = 0};
 
-#if 1
-  if (cpu == 0) {
-    /* --- now we are CPU0 */
-    clear_bss();
-    page_table_fill_inner_nodes();
-    page_table_fill_leaves();
-    WRITE_SPECIALREG(TTBR1_EL1, PHYSADDR((uint64_t)&_level1_pagetable));
-    WRITE_SPECIALREG(TTBR0_EL1, PHYSADDR((uint64_t)&_level1_pagetable));
-    enable_mmu();
+  if (cpu != 0) {
+    result = cpu_wait();
   }
-#endif
+
+  clear_bss();
```

Comment:
This and following two lines should be performed only on CPU0. Am I missing something?

Reply:
You are right. Fixed.